File: ref/bitunix-demo/Demo/Node/openApiHttpFuturePublic.js
const axios = require('axios');
const ErrorCode = require('./errorCodes');
const fs = require('fs');
const path = require('path');

class OpenApiHttpFuturePublic {
    /**
     * initialize the futures public api client
     * @param {string} configPath  config file path, default is 'config.json'
     */
    constructor(configPath = 'config.json') {
        this.loadConfig(configPath);
        this.defaultHeaders = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'language': 'en-US'
        };
    }

    /**
     * load config file
     * @param {string} configPath  config file path
     * @private
     */
    loadConfig(configPath) {
        try {
            const configContent = fs.readFileSync(path.resolve(__dirname, configPath), 'utf8');
            const config = JSON.parse(configContent);
            this.baseUrl = config.http.uri_prefix;
        } catch (error) {
            throw new Error(`failed to load config file: ${error.message}`);
        }
    }

    /**
     * handle response data
     * @param {Object} response  response data
     * @returns {Object} processed response data
     * @private
     */
    handleResponse(response) {
        const { code, msg, data } = response;
        if (code !== 0) {
            const error = ErrorCode.getByCode(code);
            if (error) {
                throw new Error(ErrorCode.getMessage(code));
            }
            throw new Error(`未知错误: ${code} - ${msg}`);
        }
        return data;
    }

    /**
     * send HTTP request
     * @param {string} method  request method
     * @param {string} url  request URL
     * @param {Object} params  request parameters
     * @returns {Promise<Object>} response data
     * @private
     */
    async sendRequest(method, url, params = {}) {
        try {
            const config = {
                method,
                url: `${this.baseUrl}${url}`,
                headers: this.defaultHeaders
            };

            if (method === 'GET') {
                config.params = params;
            } else {
                config.data = params;
            }

            const response = await axios(config);
            return this.handleResponse(response.data);
        } catch (error) {
            if (error.response) {
                throw new Error(`HTTP错误: ${error.response.status}`);
            }
            throw error;
        }
    }

    /**
     * get futures trading pair market data
     * @param {string} [symbols]  futures trading pair, multiple separated by commas, e.g. BTCUSDT,ETHUSDT
     * @returns {Promise<Object>} market data
     */
    async getTickers(symbols = null) {
        const params = {};
        if (symbols) {
            params.symbols = symbols;
        }
        return this.sendRequest('GET', '/api/v1/futures/market/tickers', params);
    }

    /**
     * get depth data
     * @param {string} symbol  futures trading pair
     * @param {number} [limit=100]  depth quantity
     * @returns {Promise<Object>} depth data
     */
    async getDepth(symbol, limit = 100) {
        const params = {
            symbol,
            limit
        };
        return this.sendRequest('GET', '/api/v1/futures/market/depth', params);
    }

    /**
     * get K-line data
     * @param {string} symbol  futures trading pair
     * @param {string} interval K-line interval, e.g. 1m, 5m, 15m, 30m, 1h, 4h, 1d
     * @param {number} [limit=100]  data quantity, default 100, max 200
     * @param {number} [startTime]  start time (Unix timestamp, millisecond format)
     * @param {number} [endTime]  end time (Unix timestamp, millisecond format)
     * @param {string} [type='LAST_PRICE'] K-line type, optional values: LAST_PRICE (latest price), MARK_PRICE (mark price)
     * @returns {Promise<Object>} K-line data
     */
    async getKline(symbol, interval, limit = 100, startTime = null, endTime = null, type = 'LAST_PRICE') {
        const params = {
            symbol,
            interval,
            limit,
            type
        };

        if (startTime) {
            params.startTime = startTime;
        }
        if (endTime) {
            params.endTime = endTime;
        }

        return this.sendRequest('GET', '/api/v1/futures/market/kline', params);
    }

    /**
     * get batch funding rate
     * @returns {Promise<Object>} funding rate data
     */
    async getBatchFundingRate() {
        const params = {};

        return this.sendRequest('GET', '/api/v1/futures/market/funding_rate/batch', params);
    }
}

// example usage
async function main() {
    try {
        // create client instance
        const client = new OpenApiHttpFuturePublic();

        // get market data
        const tickers = await client.getTickers("BTCUSDT,ETHUSDT");
        console.log("tickers data:", JSON.stringify(tickers, null, 2));

        // get depth data
        const depth = await client.getDepth("BTCUSDT", 5);
        console.log("depth data:", JSON.stringify(depth, null, 2));

        // get K-line data
        const currentTime = Date.now();
        const oneHourAgo = currentTime - (60 * 60 * 1000);
        const klines = await client.getKline(
            "BTCUSDT",
            "1m",
            5,
            oneHourAgo,
            currentTime,
            "LAST_PRICE"
        );
        console.log("K-line data:", JSON.stringify(klines, null, 2));

        // get batch funding rate
        const fundingRates = await client.getBatchFundingRate();
        console.log("funding rate data:", JSON.stringify(fundingRates, null, 2));

    } catch (error) {
        console.error("error:", error.message);
    }
}

// if this file is directly run, then execute the main function
if (require.main === module) {
    main();
}

module.exports = OpenApiHttpFuturePublic; 

File: ref/bitunix-demo/Demo/Golang/open_api_http_future_public.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// OpenApiHttpFuturePublic future public api client
type OpenApiHttpFuturePublic struct {
	baseURL string
	client  *http.Client
	config  *Config
}

// NewOpenApiHttpFuturePublic create a new future public api client
func NewOpenApiHttpFuturePublic(config *Config) *OpenApiHttpFuturePublic {
	return &OpenApiHttpFuturePublic{
		baseURL: config.HTTP.URIPrefix,
		client:  &http.Client{},
		config:  config,
	}
}

// handleResponse handle api response
func (c *OpenApiHttpFuturePublic) handleResponse(resp *http.Response) (interface{}, error) {
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP Error: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Code int         `json:"code"`
		Msg  string      `json:"msg"`
		Data interface{} `json:"data"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if result.Code != 0 {
		error := GetByCode(result.Code)
		if error != nil {
			return nil, fmt.Errorf("%s", error)
		}
		return nil, fmt.Errorf("unknown error: %d - %s", result.Code, result.Msg)
	}

	return result.Data, nil
}

// BuildQueryString build get request query string
func BuildQueryString(params map[string]string) string {
	if len(params) == 0 {
		return ""
	}

	var queryString strings.Builder
	first := true
	for k, v := range params {
		if !first {
			queryString.WriteString("&")
		}
		queryString.WriteString(k)
		queryString.WriteString("=")
		queryString.WriteString(v)
		first = false
	}
	return "?" + queryString.String()
}

// GetTickers get future trading pair market data
// Rate Limit: 10 req/sec/ip
func (c *OpenApiHttpFuturePublic) GetTickers(symbols string) ([]interface{}, error) {
	params := make(map[string]string)
	if symbols != "" {
		params["symbols"] = symbols
	}

	queryString := BuildQueryString(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/tickers"+queryString, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.([]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// GetDepth get depth data
// Rate Limit: 10 req/sec/ip
// limit: Fixed gear enumeration value: 1/5/15/50/max
func (c *OpenApiHttpFuturePublic) GetDepth(symbol string, limit string) (map[string]interface{}, error) {
	params := make(map[string]string)
	params["symbol"] = symbol
	params["limit"] = limit

	queryString := BuildQueryString(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/depth"+queryString, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// GetKline get kline data
// Rate Limit: 10 req/sec/ip
// interval: kline interval such as 1m 5m 15m 30m 1h 2h 4h 6h 8h 12h 1d 3d 1w 1M
// type: Kline type, values: LAST_PRICE, MARK_PRICE; default: LAST_PRICE
func (c *OpenApiHttpFuturePublic) GetKline(symbol, interval string, limit int, startTime, endTime *int64, klineType string) ([]interface{}, error) {
	params := make(map[string]string)
	params["symbol"] = symbol
	params["interval"] = interval
	params["limit"] = fmt.Sprintf("%d", limit)

	if klineType == "" {
		klineType = "LAST_PRICE"
	}
	params["type"] = klineType

	if startTime != nil {
		params["startTime"] = fmt.Sprintf("%d", *startTime)
	}
	if endTime != nil {
		params["endTime"] = fmt.Sprintf("%d", *endTime)
	}

	queryString := BuildQueryString(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/kline"+queryString, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.([]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// GetFundingRate get funding rate
// Rate Limit: 10 req/sec/ip
func (c *OpenApiHttpFuturePublic) GetFundingRate(symbol string) (map[string]interface{}, error) {
	params := make(map[string]string)
	params["symbol"] = symbol

	queryString := BuildQueryString(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/funding_rate"+queryString, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// GetTradingPairs get trading pair info
// Rate Limit: 10 req/sec/ip
func (c *OpenApiHttpFuturePublic) GetTradingPairs() ([]interface{}, error) {
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/trading_pairs", nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.([]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// GetBatchFundingRate get batch funding rate
// Rate Limit: 10 req/sec/ip
func (c *OpenApiHttpFuturePublic) GetBatchFundingRate() ([]interface{}, error) {
	params := make(map[string]string)

	queryString := BuildQueryString(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/market/funding_rate/batch"+queryString, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := c.handleResponse(resp)
	if err != nil {
		return nil, err
	}

	result, ok := data.([]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", data)
	}

	return result, nil
}

// HttpPublicExampleUsage example usage
func HttpPublicExampleUsage() {
	// load config
	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("load config failed: %v\n", err)
		return
	}

	// create client
	client := NewOpenApiHttpFuturePublic(config)

	// get market data
	tickers, err := client.GetTickers("BTCUSDT,ETHUSDT")
	if err != nil {
		fmt.Printf("get market data failed: %v\n", err)
		return
	}
	fmt.Printf("market data: %+v\n", tickers)

	// get depth data
	depth, err := client.GetDepth("BTCUSDT", "5")
	if err != nil {
		fmt.Printf("get depth data failed: %v\n", err)
		return
	}
	fmt.Printf("depth data: %+v\n", depth)

	// get kline data
	currentTime := time.Now().UnixMilli()
	oneHourAgo := currentTime - (60 * 60 * 1000)
	klines, err := client.GetKline("BTCUSDT", "1m", 5, &oneHourAgo, &currentTime, "LAST_PRICE")
	if err != nil {
		fmt.Printf("get kline data failed: %v\n", err)
		return
	}
	fmt.Printf("kline data: %+v\n", klines)

	// get funding rate
	fundingRate, err := client.GetFundingRate("BTCUSDT")
	if err != nil {
		fmt.Printf("get funding rate failed: %v\n", err)
		return
	}
	fmt.Printf("funding rate: %+v\n", fundingRate)

	// get batch funding rate
	fundingRates, err := client.GetBatchFundingRate()
	if err != nil {
		fmt.Printf("get batch funding rate failed: %v\n", err)
		return
	}
	fmt.Printf("funding rates: %+v\n", fundingRates)

	// get trading pair info
	tradingPairs, err := client.GetTradingPairs()
	if err != nil {
		fmt.Printf("get trading pair info failed: %v\n", err)
		return
	}
	fmt.Printf("trading pair info: %+v\n", tradingPairs)
}
